source/developer_manual/design/dict_protocol.rst

```diff
@@ -0,0 +1,251 @@
+
+.. code-block:: none
+
+   C: I<flags>TAB<max rows>TAB<path>TAB(<user>)
+
+The iteration flag is a integer with following options:
+
++----------------+--------+--------------------------------------------------+
+| flag           | value  | meaning                                          |
++================+========+==================================================+
+| RECURSE        | 0x01   | Recurse to all the sub-hierarchies               |
++----------------+--------+--------------------------------------------------+
+| SORT_BY_KEY    | 0x02   | Sort returned results by key                     |
++----------------+--------+--------------------------------------------------+
+| SORT_BY_VALUE  |  0x04  | Sort returned results by value                   |
++----------------+--------+--------------------------------------------------+
+| NO_VALUE       | 0x08   | Don't return values, only keys                   |
++----------------+--------+--------------------------------------------------+
+| EXACT_KEY      | 0x10   | Don't recurse at all. This is basically the same |
+|                |        | as LOOKUP comman dbut it will return all of the  |
+|                |        | rows instead of only the first one.              |
++----------------+--------+--------------------------------------------------+
+| FLAG_ASYNC     | 0x20   | Perform iteration asynchronously                 |
++----------------+--------+--------------------------------------------------+
+
+
+BEGIN command
+^^^^^^^^^^^^^
+
+Begins a dict transaction with the given ID.
+
+.. code-block:: none
+
+   C: B<id>TAB(<user>)
+
+Note that transactions commands don't have replies with exception of COMMIT.
+After a transaction is successfully started, transaction operations i.e.
+SET, UNSET, ATOMIC_INC, and TIMESTAMP can be used.
+
+COMMIT command
+^^^^^^^^^^^^^^
+
+Commit the transaction corresponding to the given ID.
+
+.. code-block:: none
+
+   C: C<transaction id>
+
+Note that the transaction is looked up from the ID so no username is required.
+
+COMMIT_ASYNC command
+^^^^^^^^^^^^^^^^^^^^
+
+Commit the transaction in asynchronous mode.
+
+.. code-block:: none
+
+   C: D<transaction id>
+
+The async commit is currently used by dovecot's dict client even though it is
+supported by dict server.
+
+ROLLBACK command
+^^^^^^^^^^^^^^^^
+
+Rollback changes made in the transaction.
+
+.. code-block:: none
+
+   C: R<transaction id>
+
+SET command
+^^^^^^^^^^^
+
+Set a value for the given key.
+
+.. code-block:: none
+
+   C: S<transaction id>TAB<key>TAB<value>
+
+UNSET command
+^^^^^^^^^^^^^
+
+Unset a record in the dictionary given the key.
+
+.. code-block:: none
+
+   C: U<transaction id>TAB<key>
+
+ATOMIC_INC command
+^^^^^^^^^^^^^^^^^^
+
+Atomically increment numeric value of a key with the given increment value.
+Note that the value is changed when transaction is being committed, so it's
+known beforehand what the value will become. The value is updated only
```

Comment:
it's not known?..

Reply:
Since it's deferred to COMMIT. If there are SETs after the ATOMIC_INC the final value would be different.

Reply:
yes, but the text is missing the "not" word

Reply:
Oh. Completely missed your point :) fixed.